spec/unit/hanami/assets/configuration_spec.rb

```diff
@@ -465,6 +465,68 @@
     end
   end
 
+  describe "#crossorigin?" do
+    after do
+      @configuration.reset!
+    end
+
+    context "development mode" do
+      before do
+        @configuration.load!
+      end
+
+      it "returns false when scheme, host, and port match" do
+        expect(@configuration.crossorigin?("http://localhost:2300/assets/application.js")).to be(false)
+      end
+
+      it "returns true when scheme doesn't match" do
+        expect(@configuration.crossorigin?("https://localhost:2300/assets/application.js")).to be(true)
+      end
+
+      it "returns true when host doesn't match" do
+        expect(@configuration.crossorigin?("http://some-host:2300/assets/application.js")).to be(true)
+      end
+
+      it "returns true when uses a subdomain" do
+        expect(@configuration.crossorigin?("http://assets.localhost:2300/assets/application.js")).to be(true)
+      end
+
+      it "returns true when port doesn't match" do
+        expect(@configuration.crossorigin?("http://localhost:8080/assets/application.js")).to be(true)
+      end
+    end
+
+    describe "production mode" do
+      before do
+        @configuration.scheme "https"
+        @configuration.host   "hanamirb.org"
+        @configuration.port   443
+        @configuration.load!
+      end
+
+      it "returns false when scheme, host, and port match" do
+        expect(@configuration.crossorigin?("https://hanamirb.org/assets/application.js")).to be(false)
+      end
+
+      it "returns true when scheme doesn't match" do
+        expect(@configuration.crossorigin?("http://hanamirb.org/assets/application.js")).to be(true)
+      end
+
+      it "returns true when host doesn't match" do
+        expect(@configuration.crossorigin?("https://hanamirb.test/assets/application.js")).to be(true)
+      end
+
+      it "returns true when uses a subdomain" do
+        expect(@configuration.crossorigin?("https://www.hanamirb.org/assets/application.js")).to be(true)
+      end
+
+      xit "returns true when port doesn't match" do
```

Comment:
Fix?

Reply:
This is an edge case that costs in terms of performance to get it right.

---

Quick example: the base url is `https://domain.test`, if a resource is linked as `https://domain.test:8080` it shouldn't pass the "Same-Origin" policy because the port is different.

That means our "Cross-Origin" (which is the opposite of "Same-Origin") check should return `true`, but it returns `false`.

Please notice that this check fails **only** if the asset is linked as `src="https://domain.test:8080/assets/application.js"`. If it's linked as `src="/assets/application.js"` (the standard way), it doesn't even run the check, because it's a relative path.

CDN URLs checks (`src="https://cdn.example.test/assets/application.js"`) are correct as well.

---

The current implementation isn't 100% accurate, but it's faster than one that is 100% accurate.

To get it 100% correct, for each linked asset, it should parse the URI of the assets and compare each element (`scheme`, `auth`, `host`, and `port`) with the current base URL. IMO It's too much for an edge case.

---

Given the current implementation covers the most important cases:

  * Relative URL (`/assets/application.js`)
  * Different subdomain (`https://assets.domain.test/assets/application.js`)
  * CDN (`https://cdn.example.test/assets/application.js`)

I'd keep the current implementation.